linera-chain/src/manager.rs

```diff
@@ -283,43 +283,68 @@ where
             if old_proposal.content == proposal.content {
-                return Ok(Outcome::Skip); // We already voted for this proposal; nothing to do.
+                return Ok(Outcome::Skip); // We have already seen this proposal; nothing to do.
             }
         }
         // When a block is certified, incrementing its height must succeed.
         ensure!(
             new_block.height < BlockHeight::MAX,
             ChainError::InvalidBlockHeight
         );
-        // TODO(#2971): The client still needs to make a note of proposals that fail here:
-        match self.locked.get().as_ref() {
-            None => {} // No locked block; any proposal is allowed.
-            Some(LockedBlock::Fast(old_proposal)) => {
-                // We have a locked block in the `Fast` round. Either the proposal contains a
-                // validated block certificate, or it must propose the same block.
+        let current_round = self.current_round();
+        match new_round {
+            Round::Fast => {}
+            Round::MultiLeader(_) | Round::SingleLeader(0) => {
                 ensure!(
-                    proposal.validated_block_certificate.is_some()
-                        || new_block == &old_proposal.content.block,
-                    ChainError::HasLockedBlock(new_block.height, Round::Fast)
+                    self.is_super(&proposal.owner) || !current_round.is_fast(),
+                    ChainError::WrongRound(current_round)
+                );
+                // After the fast round, proposals older than the current round are obsolete.
+                ensure!(
+                    new_round >= current_round,
+                    ChainError::InsufficientRound(new_round)
                 );
             }
-            Some(LockedBlock::Regular(validated)) => {
-                // We have a locked block in a round other than `Fast`. The proposal must contain
-                // a certificate that is no older than the locked block.
+            Round::SingleLeader(_) | Round::Validator(_) => {
+                // After the first single-leader round, only proposals from the current round are relevant.
                 ensure!(
-                    proposal
-                        .validated_block_certificate
-                        .as_ref()
-                        .is_some_and(|cert| validated.round <= cert.round),
-                    ChainError::HasLockedBlock(new_block.height, validated.round)
+                    new_round == current_round,
+                    ChainError::WrongRound(current_round)
                 );
             }
         }
+        // The round of our validation votes is only allowed to increase.
+        if let Some(vote) = self.validated_vote() {
+            ensure!(
+                new_round > vote.round,
+                ChainError::InsufficientRoundStrict(vote.round)
+            );
+        }
+        // A proposal that isn't newer than the locked block is not relevant anymore.
```

Comment:
Is this restriction just making sense or is it a safety requirement?

Reply:
It's not a safety requirement anymore, since for safety we now check `confirmed_vote` directly.